domain/application/state/application.go

```diff
@@ -683,18 +827,72 @@ DELETE FROM net_node WHERE uuid IN
 	return nil
 }
 
-func (st *ApplicationState) deleteCloudContainer(ctx context.Context, tx *sqlair.TX, unitID string) error {
-	unit := coreUnit{ID: unitID}
+func (st *ApplicationState) deleteCloudContainer(ctx context.Context, tx *sqlair.TX, netNodeID string) error {
+	cloudContainer := cloudContainer{NetNodeID: netNodeID}
+
+	if err := st.deleteCloudContainerPorts(ctx, tx, netNodeID); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := st.deleteCloudContainerAddresses(ctx, tx, netNodeID); err != nil {
+		return errors.Trace(err)
+	}
 
-	deleteCloudContainer := `DELETE FROM cloud_container WHERE net_node_uuid = $coreUnit.uuid`
-	deleteCloudContainerStmt, err := st.Prepare(deleteCloudContainer, unit)
+	deleteCloudContainerStmt, err := st.Prepare(`
+DELETE FROM cloud_container
+WHERE net_node_uuid = $cloudContainer.net_node_uuid`, cloudContainer)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	if err := tx.Query(ctx, deleteCloudContainerStmt, unit).Run(); err != nil {
+	if err := tx.Query(ctx, deleteCloudContainerStmt, cloudContainer).Run(); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (st *ApplicationState) deleteCloudContainerAddresses(ctx context.Context, tx *sqlair.TX, netNodeID string) error {
+	unit := coreUnit{
+		NetNodeID: netNodeID,
+	}
+	deleteDeviceStmt, err := st.Prepare(`
+DELETE FROM link_layer_device
+WHERE net_node_uuid = $coreUnit.net_node_uuid`, unit)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	deleteAddressStmt, err := st.Prepare(`
+DELETE FROM ip_address
+WHERE device_uuid IN
+(SELECT device_uuid FROM link_layer_device lld
+WHERE lld.net_node_uuid = $coreUnit.net_node_uuid)
+`, unit)
+	if err != nil {
 		return errors.Trace(err)
 	}
```

Comment:
Sometimes it is hard to grok these state methods because the prepares are all done first then the statement execution later. In this case the ordering of the prepares made me double take because it doesn't match the order of the statement execution. That is on me, nothing is actually wrong here (but in my head the CTE here was trying to select a device_uuid for a link_layer_device that was already deleted :sweat_smile: )

Reply:
This is a me problem, not a you problem, I'm not used to reviewing state stuff yet because I haven't written any myself.

Reply:
Fair point. I'll make the order of prepare match the order of execution.